jupyter_ydoc/ydoc.py

```diff
@@ -189,6 +201,24 @@ def set(self, value):
             self._ymeta.set(t, "nbformat", nb["nbformat"])
             self._ymeta.set(t, "nbformat_minor", nb["nbformat_minor"])
 
+    def on_cells_changed(self, event):
+        if len(event.target) == 0:
+            # We can not update a YItem from the callback
+            # create a task to do it later
+            asyncio.create_task(self.append_new_cell())
```

Comment:
Running in an event loop is a big requirement, as until now jupyter-ydoc didn't assume that.
Is there another way?

Reply:
> Is there another way?

I don't know another way

Reply:
Maybe @Waidhoferj knows?

Reply:
You should be able to append to the array inside the observe callback. The following works for me:
```python
def test_operation_inside_observe():
    d = Y.YDoc()
    array = d.get_array("array")

    def push_to_array(item):
        nonlocal array
        nonlocal d
        with d.begin_transaction() as txn:
            last_insertion = item.delta[-1]["insert"][-1]
            if len(array) < 10:
                array.append(txn, last_insertion)
            else:
                print(array)

    array.observe(push_to_array)

    with d.begin_transaction() as txn:
        array.extend(txn, ["ba", "na"])

---
Output: ['ba', 'na', 'na', 'na', 'na', 'na', 'na', 'na', 'na', 'na']
```

Reply:
Thanks John. I would really like `jupyter_ydoc` to remain a fully synchronous library, as is y-py. It is okay to have `jupyter_server_ydoc` or `ypy-websocket` rely on async, because they are inherently network-based, but `jupyter_ydoc` should be usable by other libraries in a non-async environment.

Reply:
@hbcarlos did you try what John suggested?

Reply:
Yes, and it doesn't work.

Reply:
Can you please give more details, so that we can investigate?

Reply:
```
def _on_cells_changed(self, event):
        """
        Handle a change in the cells list.

        Parameters
        ----------
            event: YArrayEvent
        """
        print("[_on_cells_changed] cells:", len(event.target))
        if len(event.target) == 0 :
            print("[_on_cells_changed] empty doc", len(self._ycells))
            ycell = self.create_ycell({
                "cell_type": "code",
                "execution_count": None,
                "metadata": {},
                "outputs": [],
                "source": "",
                "id": str(uuid4()),
            })
            print("[_on_cells_changed] ycell", ycell)

            with self._ydoc.begin_transaction() as txn:
                print("[_on_cells_changed] transaction")
                self._ycells.append(txn, ycell)
                print("[_on_cells_changed] transaction cell added", len(self._ycells))
            
            print("[_on_cells_changed] doc", len(event.target))
```

and this is the output:
```
[_on_cells_changed] cells: 0
[_on_cells_changed] empty doc 0
[_on_cells_changed] ycell {'cell_type': 'code', 'metadata': {}, 'execution_count': None, 'outputs': YArray([]), 'source': YText(), 'id': 'bc28645a-4218-47ac-b409-8c180817e4ba'}
```
There is no errors or anything.

Reply:
Thanks, I can also reproduce the issue.
@Waidhoferj it seems that nothing is executed inside the `begin_transaction()` context. Any idea?

Reply:
You should avoid transforming the document in in the event handler callback. Event handlers are fired during a transaction. In suspect that in Yrs you currently can't create another transaction within another transaction. Try manipulating the document after the callback.

@Waidhoferj the execution stopped and I suspect that Yrs fired an exception. Do you have an idea why the exception didn't bubble up to the python runtime?

Reply:
@davidbrochart Should we use the `asyncio.create_task(self.append_new_cell())` again, or do you know a better way of delaying adding a new cell?

Reply:
Have you tried [observe_after_transaction](https://ypy.readthedocs.io/en/latest/autoapi/y_py/index.html#y_py.YDoc.observe_after_transaction)?

Reply:
Yes, it doesn't work. It is the error I posted https://github.com/jupyter-server/jupyter_server_ydoc/pull/37#issue-1395813744

Reply:
this is the error:
```
Task exception was never retrieved
future: <Task finished name='Task-189' coro=<WebsocketServer.serve() done, defined at /home/carlos/micromamba/envs/rtc/lib/python3.10/site-packages/ypy_websocket/websocket_server.py:101> exception=SystemError('<built-in function apply_update> returned a result with an exception set')>
Traceback (most recent call last):
  File "/home/carlos/Documents/rtc/jupyter_ydoc/jupyter_ydoc/ydoc.py", line 254, in _on_after_transaction
    with self._ydoc.begin_transaction() as txn:
RuntimeError: Already mutably borrowed

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/carlos/micromamba/envs/rtc/lib/python3.10/site-packages/ypy_websocket/websocket_server.py", line 121, in serve
    update = await process_message(message, room.ydoc, websocket)
  File "/home/carlos/micromamba/envs/rtc/lib/python3.10/site-packages/ypy_websocket/yutils.py", line 112, in process_message
    Y.apply_update(ydoc, update)
SystemError: <built-in function apply_update> returned a result with an exception set
```

Reply:
if you want to try it out. in this PR add:
`self._ydoc.observe_after_transaction(self._on_after_transaction)` in `initialize`
and:
```python
    def _on_after_transaction(self, event):
        """
        Handle a change in the cells list.

        Parameters
        ----------
            event: YArrayEvent
        """
        print("[_on_after_transaction] cells:", len(self._ycells))
        if len(self._ycells) == 0 :
            print("[_on_after_transaction] empty doc", len(self._ycells))
            ycell = self.create_ycell({
                "cell_type": "code",
                "execution_count": None,
                "metadata": {},
                "outputs": [],
                "source": "",
                "id": str(uuid4()),
            })
            print("[_on_after_transaction] ycell", ycell)

            with self._ydoc.begin_transaction() as txn:
                self._ycells.append(txn, ycell)
            
            print("[_on_after_transaction] doc", len(self._ycells))
```

Reply:
@dmonad @Waidhoferj `observe_after_transaction` cannot be used in this case either? What are the options to mutate a document after detecting a change?

Reply:
The same is true for `observe_after_transaction` (it is fired at the end of the current transaction - but not technically after the transaction).

You could make the change directly after the callback. E.g. after a timeout (+ you should perform a sanity check first like checking the document size first). 

In general, manipulating a Y* document within (or because of) an observer call is an anti-pattern. A few reasons:

1. Another observer will be fired because of your change which can trigger yet another type of change. It can result in infinite loops.
2.  If you have two actors who "insert a new cell if the document is empty", then you can easily end up with two (or more) cells because every actor will try to "fix" the document, leading to yet another broken document.
3. Most importantly (not very obvious): You can get into unpredictable states. If you manipulate a type within an observer call - before other observers are fired - it is unclear in which order the observers for the new change will be fired. In any case, the observer for the first change can be fired even though the document has already been manipulated by change2, which often leads to inconsistencies when accessing the data using the supplied diff.

By all means, it should be avoided to rely on a central server that "cleans up" the document. It might make some things easier but will hurt in the long-run and lead to issues that are hard to debug.

A better pattern would be that the user inserts a new cell when it deletes the last cell. There is a small chance that two users delete the last cell simultaneously. Hence, you should principally allow to have an empty document. In this rare case, users can still fix the document by clicking the "new cell" button.

Reply:
> A better pattern would be that the user inserts a new cell when it deletes the last cell.

It's actually the current state, but when there are N clients, this leads to N cells being inserted, because all clients see the document as empty. That's the bug we are trying to fix.

Reply:
It is the same that I'm doing [here](https://github.com/hbcarlos/jupyterlab/blob/9fcf102a3eff94f6b54b7a8f4f637b7af6ea8fc8/packages/notebook/src/widget.ts#L590-L609). That's wrong, it works because I'm adding the cell only when we are not in collaborative mode (only one client is adding the cell). But what do you think will happens, If we, one day, decide to allow users to open a notebook twice in the same JupyterLab instance? We will have two clients reacting to, updating, and adding the cell.

Reply:
Well put @hbcarlos. The change should not happen because of a reaction to an event. I think a good place for inserting the new cell would be in the 'deleteCell' method in the shared model.